src/osd/PrimaryLogPG.cc

```diff
@@ -5453,7 +5453,7 @@ int PrimaryLogPG::do_read(OpContext *ctx, OSDOp& osd_op) {
   dout(30) << __func__ << " op.extent.truncate_size: " << op.extent.truncate_size << dendl;
 
   // are we beyond truncate_size?
-  if ( (seq < op.extent.truncate_seq) &&
+  if ( seq && (seq < op.extent.truncate_seq) &&
```

Comment:
In a quick check, it sure looks like oi.truncate_seq should start out at 0 and only be incremented when truncates actually happen. (Which, in the reproducer at https://tracker.ceph.com/issues/37378, should indeed have been done.)

This makes me think that the issue is on the filesystem side, with not setting the correct truncate_seq values, or perhaps the kernel isn't ordering OSD-level truncate_seq values correctly with the data that has actually been written out to RADOS objects. @ukernel, is that plausible?

Reply:
Thank you for your review!

One thing I forgot to mention is that if I do a write(2), overwriting the whole file, I'm able to read the data I just wrote; but doing another object copy on top will always return 0s again.  That's what made me think about an OSD bug... but it can still be an issue on the kernel side, of course.  Obviously, I spent already a lot of time looking at the kernel code (where I feel a bit more at home :-) ) but failed to figure out what was going on.

Reply:
Yeah, what I'm specifically suspecting here is that the kernel is blindly doing a truncate_seq increment on the backing objects even when the previous version of the file only ever existed in its local memory buffer, and doesn't increment the object info's truncate_seq to match.
(I don't remember exactly how that write interface works, but it's my guess from the symptoms described so far!)